### PR TITLE
Arm64

### PR DESCRIPTION
**Pull request by dmikusa**
## Summary

Support for multi-arch images, specifically amd64 and arm64.

This PR includes:

- changes to the stack & docker file so that it will build both architectures
- a script `publish.sh` to compliment `create.sh`
- changes to CI so that it will publish the multi-arch images correctly (`skopeo copy` doesn't work)

## Use Cases

ARM64 is 🌶️ 

## Questions

1. I'm not 100% sure if this really is a minor change. I don't think it will break anything. Most clients, including pack, are capable of using multi-arch images and just silently pick the arch to match the local arch. We might flag this as a major though, not because it will break things but because it's just a big change?
2. I'm not 100% sure how the CI system works. I gave it a try. If that's not right, or needs to go into another repo, let me know.
3. I assume it will start publishing multi-arch images when this PR is merged. Caution is probably warranted. Not sure if there is a way we can publish experimental images for a while first? 🤔 

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

**Comment by dmikusa** [comment]
Hmm, any ideas why [this](https://github.com/paketo-buildpacks/jammy-tiny-stack/actions/runs/4878806254/jobs/8704862211#step:5:45) is failing? 

It works locally:

```
> ./scripts/create.sh
Using jam 2.2.0
Building io.buildpacks.stacks.jammy.tiny
  Building on linux/amd64
    Building base images
      Build complete for base images
    build: Decorating base image
      Adding CNB_* environment variables
      Adding io.buildpacks.stack.* labels
      Creating cnb user
    run: Decorating base image
      Adding io.buildpacks.stack.* labels
      Creating cnb user
      Updating /etc/os-release
    build: Updating image
    run: Updating image

  Building on linux/arm64
    Building base images
      Build complete for base images
    build: Decorating base image
      Adding CNB_* environment variables
      Adding io.buildpacks.stack.* labels
      Creating cnb user
    run: Decorating base image
      Adding io.buildpacks.stack.* labels
      Creating cnb user
      Updating /etc/os-release
    build: Updating image
    run: Updating image

  Exporting build image to /Users/dmikusa/Code/OSS/paketo-buildpacks/jammy-tiny-stack/build/build.oci
  Exporting run image to /Users/dmikusa/Code/OSS/paketo-buildpacks/jammy-tiny-stack/build/run.oci
```

**Comment by sophiewigmore** [comment]
Hey Dan! I'll try to take a look at this in the next couple of days

**Comment by sophiewigmore** [comment]
Also, I agree that a major bump could be used here probably since the introduction of the arm64 stack is a pretty big change

**Comment by sophiewigmore** [comment]
btw I am running into an issue building the ARM64 stack locally as well. I was under the impression that compiling the multi-arch image on my machine would just work, but that doesn't appear to be the case.

**Comment by sophiewigmore** [comment]
I can build fine on my local machine (Mac w Intel processor) so I'm still trying to investigate what the runner env. is missing to prevent the builds from succeeding

**Comment by sophiewigmore** [comment]
@dmikusa @robdimsdale since we're running on linux here, I added a QEMU install action to the test workflow (and we will likely have to do the same for the stack creation workflow) and the build succeeded: https://github.com/paketo-buildpacks/jammy-tiny-stack/actions/runs/4940128818/jobs/8831489478

The acceptance tests themselves now fail for a different reason, but the build works at least

**Comment by robdimsdale** [comment]
Awesome, thanks for digging into this @sophiewigmore ! It makes sense that we'd need to add QEMU to enable building for different architectures. I haven't dug into whether this is the best (or perhaps even only) long-term strategy but it's certainly fine for now.

The tests are failing because we haven't updated the tests to be architecture-independent. The tests are hardcoded to check for amd64 metadata, which obviously doesn't exist on the arm64 stack. We should continue to use this PR to drive out that change to the tests, and that will enable us to use this PR as a model for other stacks.

---

Regarding my comment above about whether QEMU is the best/only path forward, I notice that the stack creation time has increased from about one minute to about five minutes. I assume that the arm64 build is running under QEMU emulation which would explain the slowdown. I could imagine that for the full stack this slowdown could be significant because it's already takes 4-5 minutes to build the stack natively.

I don't know if there's some other option that could look like building the arm64 stack natively on the amd64 runner by linking to arm64 packages and downloading arm64 binaries, rather than executing the build process under emulation, which is what I assume is happening.

cc @maliksalman as someone who might have relevant thoughts here.

**Comment by dmikusa** [comment]
Woohoo! Thanks @sophiewigmore really appreciate you getting to the bottom of this & thanks @robdimsdale for all the thoughts and feedback on this. I'll take a look at the tests and see if I can get things passing.

------

@robdimsdale I agree that QEMU is probably the best approach for now. I think it is the least impact on the workflows already being used because `jam create-stack` can build both images and generate one image file with both architectures. We can then `jam publish-stack` that image.

We have ARM64 runners and could use those to build the stack images, but the entire workflow would be different then. You would need to have jobs that run on each architecture which build individual images, push those to a registry (or build tar files & upload them as assets), and in a final job pull down both images (or download assets and load them), make an index image from images, and then `jam publish-stack` the image.

In other words, I think you'd lose some of the convenience because you have to split things across multiple runners. I agree that it would probably be faster though. You would have some overhead transferring the intermediate images around, but you'd be able to build the images in parallel and on their native architectures.

**Comment by dmikusa** [comment]
🥳 Tests are passing now! 

@sophiewigmore or @robdimsdale - If you have another sec to take a look. I think we might be all set.

**Comment by sophiewigmore** [comment]
@dmikusa sorry, I was OOO last week. I'll take a look!

File: init_test.go [hunk 1]
```diff
@@ -31,10 +31,10 @@ func TestAcceptance(t *testing.T) {
 	Expect(err).ToNot(HaveOccurred())
 
 	stack.BuildArchive = filepath.Join(root, "build", "build.oci")
-	stack.BuildImageID = fmt.Sprintf("stack-build-%s", uuid.NewString())
+	stack.BuildImageID = fmt.Sprintf("localhost:8000/stack-build-%s", uuid.NewString())
```
Comment (sophiewigmore) [inline init_test.go:34]: Why do we need this now?

File: buildpack_integration_test.go [hunk 2]
```diff
@@ -16,8 +20,11 @@ import (
 	"github.com/paketo-buildpacks/occam"
 	. "github.com/paketo-buildpacks/occam/matchers"
 	"github.com/paketo-buildpacks/packit/v2/pexec"
+	"github.com/paketo-buildpacks/packit/v2/vacation"
 )
 
+const REGISTRY_IMAGE = "registry:2"
```
Comment (sophiewigmore) [inline buildpack_integration_test.go:26]: could you explain a bit about the test changes you made here to get things passing?

**Review by sophiewigmore — APPROVED** [review]
I have a couple non-blocking questions about some of the test changes here, but overall this looks good 🔥

File: init_test.go (hunk 1 above)
Comment (dmikusa) [reply init_test.go:34]: See your next comment, they are related.

File: buildpack_integration_test.go (hunk 2 above)
Comment (dmikusa) [reply buildpack_integration_test.go:26]: This is what I mentioned in the commit comment.

>- Modify integration tests so that we don't use skopeo, which can't handle multi-arch images
>
>The latter was a little tricky. I couldn't find a way to load the multi-arch OCI TAR file into the Docker daemon so we could run the test. Instead, I had to spin up a Registry running in Docker, then load the multi-arch OCI TAR files for the stack into the internal Registry. Then we can build from the registry.

It was failing because `skopeo` doesn't work with multi-arch image. At least, it doesn't work with multi-arch images the way we are generating them.

The integration test is building a stack image & running a build with it. The critical step is taking the TAR file from the build step and loading it into the Docker daemon so the build can be run. However, we can't do that with `skopeo`.

I looked around to see if I could make the go container library load an OCI TAR file into the Daemon but it doesn't seem to be supported. It can load in a Docker TAR file, but that is not the same.

I was thinking about trying to convert them or making `jam` write out a Docker TAR file, but it ended up being the easiest way to run a local registry in Docker. The go container library can quickly load the OCI TAR into the local registry, and the build can then run by pointing to the local registry for the stack image.

So it adds a little overhead, but it was not a significant amount of additional time (loading to the Daemon vs loading to a local registry).

If the local registry is a problem, we could load the image to a remote registry. I avoided that option for two reasons: a.) I was worried about network overhead and b.) it would mean we need to publish somewhere semi-public. Let me know what you think.

Comment (sophiewigmore) [reply buildpack_integration_test.go:26]: Thank you very much for the explanation - that clears things up for me. I think this approach makes sense over pushing to an intermediate remote testing registry
